### PR TITLE
Use embeds for Package & Release metadata

### PR DESCRIPTION
**Pull request by wojtekmach**
Related to: https://github.com/hexpm/hex_web/issues/205

### CODE REVIEW

File: lib/hex_web/validation.ex [hunk 1]
```diff
@@ -52,27 +17,14 @@ defmodule HexWeb.Validation do
...
-      errors = req_errors ++ type_errors
-
-      if errors == [],
-          do: [],
-        else: [{field, errors}]
-    end)
+  @doc """
+  Adds embed errors to the changeset.errors
+  """
+  def put_embed_errors(%Ecto.Changeset{changes: changes} = changeset, embed) do
```
Comment (wojtekmach) [inline lib/hex_web/validation.ex]: I think this behavior could go into Ecto directly, the way it works right now is: `changeset.errors == []` here: https://github.com/elixir-lang/ecto/blob/v2.0.0-beta.1/test/ecto/changeset/embedded_test.exs#L93

File: test/controllers/api/release_controller_test.exs [hunk 2]
```diff
@@ -87,7 +87,7 @@ defmodule HexWeb.API.ReleaseControllerTest do
     assert conn.status == 422
     body = Poison.decode!(conn.resp_body)
     assert body["message"] == "Validation error(s)"
-    assert body["errors"]["meta"]["links"] == "expected type dict(string, string)"
+    assert body["errors"]["meta"]["links"] == "is invalid"
```
Comment (wojtekmach) [inline test/controllers/api/release_controller_test.exs]: it looks like by using embeds it's relying on ecto "type-checking" and so the errors are always "is invalid" when there's a type mismatch. Not sure if there's a way to overwrite this behaviour. Perhaps it's good enough to have it like this?

File: web/models/audit_log.ex [hunk 3]
```diff
@@ -22,10 +22,19 @@ defmodule HexWeb.AuditLog do
...
+  defp do_serialize(schema), do: Map.take(schema, fields(schema))
 
   defp fields(%HexWeb.Key{}), do: [:id, :name]
-  defp fields(%HexWeb.Package{}), do: [:id, :name, :meta]
-  defp fields(%HexWeb.Release{}), do: [:id, :version, :checksum, :meta, :has_docs, :package_id]
+  defp fields(%HexWeb.Package{}), do: [:id, :name]
+  defp fields(%HexWeb.Release{}), do: [:id, :version, :checksum, :has_docs, :package_id]
   defp fields(%HexWeb.User{}), do: [:id, :username, :email, :confirmed]
+  defp fields(%HexWeb.PackageMetadata{}), do: HexWeb.PackageMetadata.__schema__(:fields)
+  defp fields(%HexWeb.ReleaseMetadata{}), do: HexWeb.ReleaseMetadata.__schema__(:fields)
```
Comment (wojtekmach) [inline web/models/audit_log.ex]: this reflection feels a bit wrong, and the same thing is used below in the View to get all fields from metadata into JSON response. Perhaps `@derive {Poison.Encoder, ...}` would be better in this particular case?

File: web/models/package_metadata.ex [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    field :contributors, {:array, :string}
+  end
+
+  @required_fields ~w(description)
+  @optional_fields ~w(licenses links maintainers)
+
+  def changeset(meta, params \\ :empty) do
+    cast(meta, params, @required_fields, @optional_fields)
+    |> validate_presence(:description)
+  end
```
Comment (wojtekmach) [inline web/models/package_metadata.ex]: for now, I tried to only port existing validations into this {Package,Release}Metadata - I'm sure more could be added

File: web/views/api/package_view.ex [hunk 5]
```diff
@@ -9,7 +9,8 @@ defmodule HexWeb.API.PackageView do
 
   def render("package", %{package: package}) do
     package
-    |> Map.take([:name, :meta, :inserted_at, :updated_at])
+    |> Map.take([:name, :inserted_at, :updated_at])
+    |> Map.put(:meta, Map.take(package.meta, PackageMetadata.__schema__(:fields)))
```
Comment (wojtekmach) [inline web/views/api/package_view.ex]: see comment above in the AuditLog

File: test/controllers/api/release_controller_test.exs (hunk 2 above)
Comment (ericmj) [inline test/controllers/api/release_controller_test.exs]: We can leave as-is for now, but this could be useful to have in ecto. I see no reason why ecto can't include the expected type.

@josevalim Thoughts?

File: web/models/audit_log.ex (hunk 3 above)
Comment (ericmj) [inline web/models/audit_log.ex]: I think this is fine actually, declaring the fields explicitly would also be fine.

File: web/models/package_metadata.ex (hunk 4 above)
Comment (ericmj) [inline web/models/package_metadata.ex]: Sounds good.

**Comment by ericmj** [comment]
This looks good to me. Waiting for input from @josevalim, specifically about the potential ecto changes.

File: lib/hex_web/validation.ex (hunk 1 above)
Comment (josevalim) [inline lib/hex_web/validation.ex]: The idea is to keep all errors in the changesets and use the `Ecto.Changeset.traverse_errors` functions to traverse them into a single list.

File: web/models/audit_log.ex (hunk 3 above)
Comment (josevalim) [inline web/models/audit_log.ex]: :+1: for declaring the fields explicitly. We do not necessarily want to include them all.

File: web/models/package_metadata.ex [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    field :maintainers, {:array, :string}
+    field :contributors, {:array, :string}
+  end
+
+  @required_fields ~w(description)
+  @optional_fields ~w(licenses links maintainers)
+
+  def changeset(meta, params \\ :empty) do
+    cast(meta, params, @required_fields, @optional_fields)
+    |> validate_presence(:description)
```
Comment (josevalim) [inline web/models/package_metadata.ex]: This will be `validate_required` once we move to Ecto 2.0. Maybe we should leave a TODO note?

File: lib/hex_web/validation.ex (hunk 1 above)
Comment (wojtekmach) [inline lib/hex_web/validation.ex]: yeah, I already started looking into `traverse_errors` to be used by
`validation_failed(conn, errors)` helper internally, but I also noticed
there's some existing code that expected `errors[:meta]` to be a list, e.g.
here:
https://github.com/hexpm/hex_web/pull/258/files#diff-c3cdcf0a302e924598045042d82d28cdR42.
Do you have a preference how this should be handled? I could leave this
`put_embed_error` helper, or I could use traverse everywhere including
tests, same as in
https://github.com/elixir-lang/ecto/blob/v2.0.0-beta.1/test/ecto/changeset/embedded_test.exs#L705

2016-03-06 10:11 GMT+01:00 José Valim <notifications@github.com>:

> In lib/hex_web/validation.ex
> <https://github.com/hexpm/hex_web/pull/258#discussion_r55135565>:
>
> > -            []
> > -          else
> > -            [{field, :missing}]
> > -          end
> > -        end)
> > -
> > -      errors = req_errors ++ type_errors
> > -
> > -      if errors == [],
> > -          do: [],
> > -        else: [{field, errors}]
> > -    end)
> > +  @doc """
> > +  Adds embed errors to the changeset.errors
> > +  """
> > +  def put_embed_errors(%Ecto.Changeset{changes: changes} = changeset, embed) do
>
> The idea is to keep all errors in the changesets and use the
> Ecto.Changeset.traverse_errors functions to traverse them into a single
> list.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hexpm/hex_web/pull/258/files#r55135565>.
>



-- 
Wojtek Mach

File: web/models/audit_log.ex (hunk 3 above)
Comment (wojtekmach) [inline web/models/audit_log.ex]: Sounds good, I'll declare the fields explicitly in both cases (here and in the view)

File: web/models/package_metadata.ex (hunk 6 above)
Comment (wojtekmach) [inline web/models/package_metadata.ex]: Nice, I'll leave a note

File: lib/hex_web/validation.ex (hunk 1 above)
Comment (josevalim) [inline lib/hex_web/validation.ex]: It is fine with the current mechanism. Maybe just leave a note on why you are unifying the errors then?

File: lib/hex_web/validation.ex [hunk 7]
```diff
@@ -52,27 +17,14 @@ defmodule HexWeb.Validation do
...
+    if changes[embed] && length(changes[embed].errors) > 0 do
```
Comment (josevalim) [inline lib/hex_web/validation.ex]: `changets[embed].errors != []` is going to be much faster!

**Comment by wojtekmach** [comment]
@ericmj @josevalim thanks for the feedback, I pushed the changes. With the change about explicitly enumerating metadata fields, I also made it explicit in the test that all of them are being saved (even if `nil`). If it's good to go, I'll squash.

File: test/controllers/api/release_controller_test.exs (hunk 2 above)
Comment (ericmj) [inline test/controllers/api/release_controller_test.exs]: I've opened an ecto issue for this.

**Comment by ericmj** [comment]
:heart:
